util/ibex_config.py

```diff
@@ -127,19 +127,82 @@ def get_config_dicts(config_file):
     return config_yaml
 
 
-def _config_dict_to_fusesoc_opts(config_dict):
-    fusesoc_cmd = []
-    for parameter, value in config_dict.items():
-        if isinstance(value, bool):
-            # For fusesoc boolean parameter are set to true if given on the
-            # command line otherwise false, it doesn't support an explicit
-            # --param=True style
-            if value:
-                fusesoc_cmd.append(shlex.quote('--' + parameter))
+class FusesocOpts:
+    def setup_args(self, arg_subparser):
+        output_argparser = arg_subparser.add_parser(
+            'fusesoc_opts', help=('Outputs options for fusesoc'))
+        output_argparser.set_defaults(output_fn=self.output)
+
+    def output(self, config_dict, args):
+        fusesoc_cmd = []
+        for parameter, value in config_dict.items():
+            if isinstance(value, bool):
+                # For fusesoc boolean parameter are set to true if given on the
+                # command line otherwise false, it doesn't support an explicit
+                # --param=True style
+                if value:
+                    fusesoc_cmd.append(shlex.quote('--' + parameter))
+            else:
+                fusesoc_cmd.append(
+                    shlex.quote('--' + parameter + '=' + str(value)))
+
+        return ' '.join(fusesoc_cmd)
+
+
+class SimOpts:
+    def __init__(self, cmd_name, description, param_set_fn, define_set_fn,
+                 hierarchy_sep):
+        self.cmd_name = cmd_name
+        self.description = description
+        self.param_set_fn = param_set_fn
+        self.define_set_fn = define_set_fn
+        self.hierarchy_sep = hierarchy_sep
+
+    def setup_args(self, arg_subparser):
+        output_argparser = arg_subparser.add_parser(
+            self.cmd_name,
+            help=('Outputs options for {0}'.format(self.description)))
+
+        output_argparser.add_argument(
+            '--ins_hier_path',
+            help=('Hierarchical path to the instance to set '
+                  'configuration parameters on'),
+            default='')
+        output_argparser.add_argument(
+            '--string_define_prefix',
+            help=('Prefix to add to defines that are used to '
+                  'pass string parameters'),
+            default='')
+        output_argparser.set_defaults(output_fn=self.output)
+
+    def output(self, config_dict, args):
+        if (args.ins_hier_path != ''):
+            ins_hier_path = args.ins_hier_path + self.hierarchy_sep
         else:
-            fusesoc_cmd.append(shlex.quote('--' + parameter + '=' + str(value)))
+            ins_hier_path = ''
+
+        sim_opts = []
+
+        for parameter, value in config_dict.items():
+            if isinstance(value, str):
+                parameter_define = args.string_define_prefix + parameter
+                define_set_str = self.define_set_fn(parameter_define, value)
+
+                if define_set_str:
```

Comment:
I think this is always true with the code as it stands, so the `else` branch is dead code. Remove the check?

Reply:
riviera_sim_opts gives None for the define string and riviera_compile_opts gives None for the parameter string so this is used.

Initially I just returned an empty string to avoid this handling but sadly the `shlex.quote` turns the empty string into a literal ''.